x-pack/plugins/apm/common/ml_job_constants.ts

```diff
@@ -16,3 +16,7 @@ export function getMlJobId(serviceName: string, transactionType?: string) {
 export function getMlIndex(serviceName: string, transactionType?: string) {
   return `.ml-anomalies-${getMlJobId(serviceName, transactionType)}`;
 }
+
+export function encodeForMlApi(value: string) {
+  return value.replace(/\s+/g, '_').toLowerCase();
```

Comment:
What about using `encodeURIComponent`? With this change `service a` and `service_a` are now the same (although it's very much an edge case).

Btw. A comment would probably be useful later (eg. "ML doesn't allow spaces in `group` properties")

Reply:
Btw. are there other characters that we should be aware of that ML doesn't support?

Reply:
I didn't consider that! But unfortunately `%` is not supported either. Not sure how we could solve this issue. See https://github.com/elastic/elasticsearch/blob/95a7eed9aa35f47b228e402508709b5bd6703cf4/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlStrings.java#L20-L26 for the regex.

Reply:
Ah, okay. Not sure how to fix then. Let's leave it for now.

Reply:
This is the function we use to validate an ML job ID in the UI.
https://github.com/elastic/kibana/blob/0b8cd6eb46d8fbc63be4d5fe29cd59c4b7b9ce7c/x-pack/plugins/ml/common/util/job_utils.js#L257

Reply:
Thanks for chiming in @jgowdyelastic.

We use a slightly more liberal regex: `^[a-zA-Z0-9 _-]+$` (alphanumeric characters, spaces, underscores, and dashes).
https://www.elastic.co/guide/en/apm/get-started/current/agents.html#choose-service-name

> // it must also start and end with an alphanumeric character'
Sounds like we'll might run into problems with service names that start or end with non-alphanumeric character eg. `service-a-` or `_internal_service`. 

What's the background for this limitation?

Reply:
There are a few reasons why we don't allow non-alphanumeric chars at the start and end of the ID. They are mainly to do with the ID being used as part of our endpoint URIs. 
We follow elasticsearch's naming convention and do now allow  `-` or `_` a the start. `.` is use for "hidden" indices etc.
Some url parsing systems don't like `.` chars at the end, e.g. slack will not auto add it to a link.
It's reasons like this why we've gone with fairly this strict rule.

Reply:
Makes sense. I don't anticipate that this will be common in service names either but currently we do allow it meaning we'll have to handle it somehow.

The easy solution would be to strip non-alphanumeric chars at the beginning and end, but that would create the same problem as I mentioned before, where `_service-A` and `service-A` are now identical (very edge casey as well)